### PR TITLE
docs: clarify usage with laravel sanctum, login / logout mutations

### PR DESCRIPTION
**Pull request by dlindenkreuz**
- [ ] Added or updated tests
- [x] Documented user facing changes
- [ ] Updated CHANGELOG.md

**Changes**

It's not obvious how to use [Laravel Sanctum](https://laravel.com/docs/7.x/sanctum) as a guard. This PR adds docs to clarify usage. Additionally, this PR covers login / logout mutations with an example, also in conjunction with Laravel Sanctum.

As discussed in #1406.

**Breaking changes**

none

### CODE REVIEW

File: docs/master/security/authentication.md [hunk 1]
```diff
@@ -45,12 +45,31 @@ This setting is used whenever Lighthouse looks for an authenticated user, for ex
...
+        // ...
+        'middleware' => [
+            \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
+            // ... other middlewares
+        ]             
+    ],
     'guard' => 'sanctum',
 ```
 
+Note that Sanctum requires you to send an CSRF token as [header](https://laravel.com/docs/7.x/csrf#csrf-x-csrf-token)
```
Comment (spawnia) [inline docs/master/security/authentication.md:65]:
```suggestion
Note that Sanctum requires you to send an CSRF token as [header](https://laravel.com/docs/csrf#csrf-x-csrf-token)
```

**Comment by spawnia** [comment]
@dlindenkreuz i will try out the suggested changes in https://github.com/nuwave/lighthouse-example and review once i find the time.

**Comment by ODelibalta** [comment]
Thank you for this, sincerely. 

But 

I did the changes based on the proposed doc waiting for review and I get page 419 page expired for my login mutation. It is a fresh install

**Comment by ODelibalta** [comment]
I've done my login according to the updated doc. It does the login mutation fine but then the default ```User``` query returns ```null``` or if I add the ```@guard``` directive to the hello world query it says unauthenticated. 

When I inspect in my browser, the X-CSRF-COOKIE  in response headers is changed on each request.

I made a button that fires a guarded hello world query and on each click lighthouse returns a different cookie value even though in my header I am sending the first cookie value I get in login page. 

```
type Query {
    users: [User!]! @paginate(defaultCount: 10)
    user(id: ID @eq): User @find
    hello: String! @guard
    me: User @auth
}

type Mutation {
    login(email: String!, password: String!): User @field(resolver: "App\\GraphQL\\Mutations\\Auth\\Login")
}

type User {
    id: ID!
    name: String!
    email: String!
    created_at: DateTime!
    updated_at: DateTime!
}
```
login resolver
```
class Login
{
    /**
     * @param  null  $_
     * @param  array<string, mixed>  $args
     */
    public function __invoke($_, array $args)
    {
        // $guard = Auth::guard();
        // When using Laravel Sanctum, always use Sanctum's guard to create a session:
        $guard = Auth::guard(config('sanctum.guard', 'web'));

        if ($guard->user()) {
            return $guard->user();
        }

        $credentials = collect($args)->only('email', 'password')->all();

        throw_unless(
            $guard->attempt($credentials),
            AuthenticationException::class
        );

        return $guard->user();
    }
}
```
cors config
```
return [ 

    'paths' => ['api/*',  'graphql', 'sanctum/csrf-cookie'],

    'allowed_methods' => ['*'],

    'allowed_origins' => ['*'],

    'allowed_origins_patterns' => [],

    'allowed_headers' => ['*'],

    'exposed_headers' => [],

    'max_age' => 0,

    'supports_credentials' => true,

];
```
client side login page
```
useEffect(() => {
    console.log(`${server}/sanctum/csrf-cookie`);
    axios.get(`${server}/sanctum/csrf-cookie`);
  }, []);

  const [loginUser, { loading }] = useMutation(LOGIN_USER, {
    errorPolicy: 'all',
    onCompleted: (res) => {
      if (res.login) {
        Router.push('/');
      } else {
        setState((prevState) => ({
          ...prevState,
          loginError: true,
        }));
      }
    },
  });
```

and in my apollo client 
```
export default function createApolloClient(initialState, ctx) {
  // The `ctx` (NextPageContext) will only be present on the server.
  // use it to extract auth headers (ctx.req) or similar.

  const authLink = setContext((_, { headers  }) => {
    // get the authentication token from local storage if it exists
    const token = Cookies.get("XSRF-TOKEN");
    // return the headers to the context so httpLink can read them
   
    return {
      headers: {
        ...headers,
        "Access-Control-Allow-Credentials": true,
        ...(token ? { "X-CSRF-TOKEN": `${token}` } : {}),
        ...(token ? { "authorization": `Bearer ${token}` } : {}),
      },
    };
  });

  const httpLink = createHttpLink({
    uri: serverUrl,
    credentials: 'same-origin',
  });

  return new ApolloClient({
    ssrMode: Boolean(ctx),
    link: authLink.concat(httpLink),
    connectToDevTools: true,
    cache: new InMemoryCache().restore(initialState),
  });
}
```

at this point I am not sure how the login works but hell world does not. I get the cookie via axios first then do login but with the same cookie I can not do a guarded hello world.

**Comment by ODelibalta** [comment]
I've implemented this based on the docs here and I can login but, when I try to access a guarded query after, it says I am unauthenticated. I inspected my browser and I am sending the cookies and I have tried the same thing with graphql playground with headers added.

**Comment by spawnia** [comment]
There is substantial amounts of setup that needs to be in place when starting from a bare-bones GraphQL only configuration.

Can you give me some feedback on https://github.com/nuwave/lighthouse-example/pull/9, how can that be improved?

**Comment by spawnia** [comment]
Thanks @dlindenkreuz

**Pull request merged**